### ISSUE TITLE
Feature: nested validation

### ISSUE THREAD

**Issue by ThatOneBro** [honojs/hono#584]
In #563, we talked about nested validation with arrays. @usualoma mentioned an interesting way to handle nested validations like this:

```ts
describe('validate nested structure by asArray', () => {
  const app = new Hono()
  app.post(
    '/post',
    validator((v) => ({
      posts: v.json('posts').asArray((nest) => ({
        title: nest.json('title').isAlpha(),
        tags: nest.json('tags').asArray().isRequired(),
        ids: nest.json('ids').asNumber().asArray(),
      }))
    })),
    (c) => {
      const res = c.req.valid()
      return c.json({ tag1: res.posts[0].tags[0] })
    }
  )

  it('Should return 200 response', async () => {
    const json = {
      post: {
        title: 'foo',
        tags: ['Workers', 'Deno', 'Bun'],
        ids: [1, 3, 5],
      },
    }
    const req = new Request('http://localhost/post', {
      method: 'POST',
      body: JSON.stringify(json),
    })
    const res = await app.request(req)
    expect(res.status).toBe(200)
    expect(await res.json()).toEqual({ tag1: 'Workers' })
  })
})
```

I think we can assign a type to this.

<img width="558" alt="image" src="https://user-images.githubusercontent.com/30598/193165303-dbf14b39-1d97-429b-8874-07ccfb6e27eb.png">

<img width="625" alt="image" src="https://user-images.githubusercontent.com/30598/193165323-95a14dff-9e94-4c05-95ca-564d958ed1b3.png">

<img width="645" alt="image" src="https://user-images.githubusercontent.com/30598/193165344-924cba2b-1ed8-4e45-a69c-dbc0d1c9c968.png">

_Originally posted by @usualoma in https://github.com/honojs/hono/issues/563#issuecomment-1262975936_

I think this coupled with some extension of the `asObject()` method, would allow clean and proper validation of any complex valid JSON structure! Hono users could then do things like this:
```ts
    validator((v) => ({
      posts: v.json('posts').asArray((nest) => ({
        title: nest.json('title').isAlpha(),
        meta: nest.json('meta').asObject((nest) => ({
          // A user should probably not be able to create posts like this, but it's an example ;)
          relatedPosts: nest.json('relatedPosts').asArray((nest) => ({
            title: nest.json('title').isAlpha(),
            tags: nest.json('tags').asArray().isRequired(),
            ids: nest.json('ids').asNumber().asArray(),
          })),
          moreStuff: nest.json('moreStuff').asArray().isOptional(),
        })),
        tags: nest.json('tags').asArray().isRequired(),
        ids: nest.json('ids').asNumber().asArray(),
      }))
    })),
    (c) => {
      const res = c.req.valid()
      return c.json({ relatedPosts1Tags: res.posts[0].meta.relatedPosts[0].tags })
    }
  )
```

This would also allow for us to return errors specific to particular paths with arrays like:
`posts[1].tags[0] is invalid - true - rule: asString` or something similar. These errors should be useful to the API consumer, and the components such as `ruleName`, `path` and `pathValue` should be exposed to custom errors (I know @yusukebe is working on a new custom validation error system). This is probably a separate issue though, and would be dependent on this issue first.

What do you all think about this plan?

**Comment by ThatOneBro** [comment]
After #592 has been merged, I would consider this issue completed. Good job @yusukebe ! 👍
